UserTopic: Don't raise errors on duplicate requests.

prakhar1144 opened the pull request:
This is a follow-up PR to #24236 ([comment](https://github.com/zulip/zulip/pull/24236#issuecomment-1457423519))

CZO discussion: [api design > duplicate request to update UserTopic row](https://chat.zulip.org/#narrow/stream/378-api-design/topic/duplicate.20request.20to.20update.20UserTopic.20row/near/1519868) 

First commit includes the changes to:
* Not raise error in the case of duplicate requests to update the visibility_policy.
* Don't send duplicate event in this case.
* Docuumented API changes.

Second commit includes the changes to: ([Discussion](https://chat.zulip.org/#narrow/stream/378-api-design/topic/duplicate.20request.20to.20update.20UserTopic.20row/near/1523806))
* Not raise error when the request is to delete a UserTopic row and the user doesn't already have a visibility_policy for the topic yet (possibly a duplicate request).
* Don't send duplicate event in such case.
* Documented API changes.

Fixes: follow-up PR to #24236

---
### Reason for decrease in db query count:
**In 'zerver/lib/user_topics.py'**

Earlier we were using `UserTopic.objects.get().delete()`, which accounts for 2 queries.
Now, we are using `UserTopic.objects.filter.delete()`, which accounts for 1 query.

### Technical Choice
* To avoid duplicate event send ( in the case of duplicate requests), I have used the approach of --`set_user_topic_visibility_policy_in_database` returning a bool variable named `skip_send_event`, using which we determine - 'whether to send event or skip'. **Let me know, if there is some better way to handle this or it is fine.**
<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,13 @@ format used by the Zulip server that they are interacting with.
+* [`PATCH /users/me/subscriptions/muted_topics`](/api/mute-topic):
+  No error is raised in the case of requests to mute an already
+  muted topic. It is an easy success.
+  Previously, the error `Topic is already muted` was raised.
```
We try to keep these brief and comprehensible without knowing anything about the endpoint or having specific strings. So I'd write "Trying to mute a topic that is already muted now results in a success response, rather than an error."

timabbott commented on zerver/lib/user_topics.py:
```diff
@@ -117,8 +117,9 @@ def set_user_topic_visibility_policy_in_database(
-    ignore_duplicate: bool = False,
-) -> None:
+) -> bool:
+    skip_send_event: bool = False
```
I think this would be better expressed with having a comment explaining that this function returns True if it modified the database, and False otherwise. The caller interprets that as whether to send an event or not, but I think logically that's what this function is telling you.

timabbott commented on zerver/actions/user_topics.py:
```diff
@@ -19,22 +19,25 @@ def do_set_user_topic_visibility_policy(
+    # If there are duplicate requests to update the visibility_policy,
+    # don't send duplicate events.
+    if skip_send_event:
+        return
```
As noted above, I think this will be a lot more readable if we have the above lines be `database_changed = set_user_topic_visibility_policy_in_database(...)` and then this is `if not database_changed: return` -- since that expresses clearly the logic that we're only sending an event if we changed something.

(And then I think you don't need the comment explaining the reasoning)

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -8202,6 +8202,10 @@ paths:
+
+        **Changes**: New in Zulip 7.0 (feature level 168), No error is raised
+        when an `add` operation is requested for a topic that has already been muted.
+        Previously, the error `Topic is already muted` was raised.
```
"error is raised" is normal language for Python code, but in the API context, we're not "raising an error/exception", we're returning an HTTP 200 success response or an HTTP 400 error. 

Also the "New in Zulip 7.0" language should only be used if the feature is new in that version -- which it isn't, we're just documenting a small change.

So this should say something like "**Changes**. Before Zulip 7.0 (feature level 168), this endpoint returned an error if asked to mute a topic that was already muted.".

The string isn't worth mentioning, because the string is tagged for translation, so a correct client shouldn't be inspecting the string at all; we use `code` for errors that we want to be parsed by clients.

timabbott commented on zerver/tests/test_user_topics.py:
```diff
@@ -114,11 +114,11 @@ def test_add_muted_topic(self) -> None:
         data["topic"] = "VERONA3"
         result = self.api_patch(user, url, data)
-        self.assert_json_error(result, "Topic already muted")
+        self.assert_json_success(result)
```
Given the comment about case-sensitivity, we should really extend this test to verify that we didn't end up with duplicate UserTopic rows with the two different cases.

prakhar1144 replied:
[Extended](https://github.com/zulip/zulip/pull/24690/commits/d5d5cbe378abe3fd3f1c14c366dd60a7ce6594c0#diff-dca29140ac4182499e88c086202b897cca41c670b6163288b76f582569528402R136) the test to verify.

abhijeetbodas2001 commented on zerver/lib/user_topics.py:
```diff
@@ -147,26 +148,19 @@ def set_user_topic_visibility_policy_in_database(
-        return
+    if duplicate_request:
+        skip_send_event = True
+        return skip_send_event
```
For this case, we may want to log (https://chat.zulip.org/#narrow/stream/378-api-design/topic/duplicate.20request.20to.20update.20UserTopic.20row/near/1522640)

prakhar1144 commented:
@timabbott This is ready to review.

**Updates in the recent push**
* Addressed the review comments.
* Added `logging.info()` and `assertLogs()` -- as [discussed](https://github.com/zulip/zulip/pull/24690#discussion_r1136524368).

**Needs your attention**
* wordings of the `logging.info` -- [first](https://github.com/zulip/zulip/pull/24690/commits/d5d5cbe378abe3fd3f1c14c366dd60a7ce6594c0#diff-97e4a396cdf669d3e133f04bc7fe98e0057cc6eece60ef2aa80ea3fd40a9eb3bR158), [second](https://github.com/zulip/zulip/pull/24690/commits/7f4fed223c111b6a0ec20acea75f5bb47d55a5d9#diff-97e4a396cdf669d3e133f04bc7fe98e0057cc6eece60ef2aa80ea3fd40a9eb3bR131) (Let me know if they need improvements)

timabbott commented on zerver/lib/user_topics.py:
```diff
@@ -117,8 +118,10 @@ def set_user_topic_visibility_policy_in_database(
-) -> None:
+) -> bool:
+    # returns True if it modifies the database and False otherwise.
+    database_changed: bool = True
```
Don't define this variable -- it's never usefully mutated. We should just `return True` / `return False` where needed.

timabbott commented on zerver/lib/user_topics.py:
```diff
@@ -147,26 +150,20 @@ def set_user_topic_visibility_policy_in_database(
-    if duplicate_request and ignore_duplicate:
-        return
+    if duplicate_request:
+        logging.info("Request to set the visibility_policy to its current value.")
```
This log line needs some details to be useful. I think `User {user_id} tried to set visibility_policy to its current value of {visibility_policy}` -- then we'll at least be able to observe trends in which users and/or which values are seeing this happen to motivate an investigation.

prakhar1144 replied:
I added a similar log for `UserTopic.VisibilityPolicy.INHERIT` case:
```python
    logging.info(
        "User %s tried to remove visibility_policy, which actually doesn't exist",
        user_profile.id,
    )
 ```

timabbott commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,12 @@ format used by the Zulip server that they are interacting with.
+* [`PATCH /users/me/subscriptions/muted_topics`](/api/mute-topic):
+  Trying to unmute a topic that was not previously muted results
+  in a success response rather than an error.
+
```
I think you messed up merge conflict resolution -- you now are adding two copies of this notice.

timabbott replied:
Oh, no, it's just very similar. I think we should put these both in feature level 169 as a change made together.

prakhar1144 replied:
Squashed the commits.

ps: This paragraph of zulip docs was really helpful in this case :)
> Overly fine commits are easy to squash later, but not vice versa. So err toward small commits, and the code reviewer can advise on squashing.

timabbott commented on zerver/lib/user_topics.py:
```diff
@@ -123,17 +121,16 @@ def set_user_topic_visibility_policy_in_database(
+            stream_id=stream_id,
+            topic_name__iexact=topic_name,
+        ).delete()
+        if deleted[0] == 0:
```
I prefer `(num_deleted, ignored) = UserTopic.objects...` as a pattern if something returns a tuple.

prakhar1144 replied:
I will update this.

I just [followed](https://github.com/zulip/zulip/blob/main/zilencer/views.py#L200) the existing pattern in the codebase.

timabbott replied:
This pattern is definitely nicer, so it wouldn't be a bad idea to update existing code using that pattern in a little follow-up PR.

prakhar1144 replied:
@timabbott 

Here's the follow-up PR #24737 

ps: I have verified that there is no more existing code in the codebase that doesn't follow this pattern.

prakhar1144 commented:
@timabbott It is ready to review.

**Updates in the recent push**
* Worked on the latest (last 4) review comments.

PIG208 commented:
@timabbott A feature level bump is missing
